src/cmap/auth/gssapi.ts

```diff
@@ -30,60 +52,161 @@ export class GSSAPI extends AuthProvider {
-          payload,
-          autoAuthorize: 1
-        };
-
-        connection.command('$external.$cmd', command, (err, result) => {
+        context!.client.step('', (err, payload) => {
           if (err) return callback(err, false);
 
-          const doc = result.result;
-          authProcess.transition(doc.payload, (err, payload) => {
-            if (err) return callback(err, false);
-            const command = {
-              saslContinue: 1,
-              conversationId: doc.conversationId,
-              payload
-            };
-
-            connection.command('$external.$cmd', command, (err, result) => {
+          externalCommand(
+            {
+              saslStart: 1,
+              mechanism: 'GSSAPI',
+              payload,
+              autoAuthorize: 1
+            },
+            (err, result) => {
               if (err) return callback(err, false);
 
               const doc = result.result;
-              authProcess.transition(doc.payload, (err, payload) => {
+              negotiate(context!, doc.payload, (err, payload) => {
                 if (err) return callback(err, false);
-                const command = {
-                  saslContinue: 1,
-                  conversationId: doc.conversationId,
-                  payload
-                };
-
-                connection.command('$external.$cmd', command, (err, result) => {
-                  if (err) return callback(err, false);
-
-                  const response = result.result;
-                  authProcess.transition(null, err => {
-                    if (err) return callback(err);
-                    callback(undefined, response);
-                  });
-                });
+
+                externalCommand(
+                  {
+                    saslContinue: 1,
+                    conversationId: doc.conversationId,
+                    payload
+                  },
+                  (err, result) => {
+                    if (err) return callback(err, false);
+
+                    const doc = result.result;
+                    finalize(context!, doc.payload, (err, payload) => {
+                      if (err) return callback(err, false);
+
+                      externalCommand(
+                        {
+                          saslContinue: 1,
+                          conversationId: doc.conversationId,
+                          payload
+                        },
+                        (err, result) => {
+                          if (err) return callback(err, false);
+
+                          callback(undefined, result.result);
+                        }
+                      );
+                    });
+                  }
+                );
               });
-            });
-          });
+            }
+          );
         });
-      });
+      }
+    );
+  }
+}
+
+function initialize(
+  context: Omit<GSSAPIContext, 'client'>,
+  callback: (err?: Error | null, context?: GSSAPIContext | null) => void
```

Comment:
Same thing here, about not use `null` for callback values.
If possible can `context` not be `null` too just undefined?
I'd personally rather:
`GSSAPIContext | undefined` and not
`GSSAPIContext | null | undefined`